aws-amplify/amplify-ios
# fix(DataStore): log metrics for ReconcileAndLocalSaveOperation

*Issue #, if available:*

*Description of changes:*
The goal of this PR is to add the ability to gather metrics on `ReconcileAndLocalSaveOperation` which performs a number of read and writes to the the local SQL DB. 

The debug logs produced look like this
```
2021-05-11 14:44:50.677372-0700 DataStoreLoadTest[77234:1618471] [ReconcileAndLocalSaveOperation] query local metadata: 0.000369192s
2021-05-11 14:44:50.677901-0700 DataStoreLoadTest[77234:1618471] [ReconcileAndLocalSaveOperation] query pending mutations: 0.000528688s
2021-05-11 14:44:50.679721-0700 DataStoreLoadTest[77234:1618471] [ReconcileAndLocalSaveOperation] save model: 0.001818936s
2021-05-11 14:44:50.681961-0700 DataStoreLoadTest[77234:1618471] [ReconcileAndLocalSaveOperation] save metadata: 0.002241446s
2021-05-11 14:44:50.682496-0700 DataStoreLoadTest[77234:1618471] [ReconcileAndLocalSaveOperation] total time: 0.005473928s
```

Testing done based on these metrics:
- records to sync 10k
- page size to sync 1k
- item size (~250B)
- a single model
- IPhone 12 Pro simulator

Results:
- query local metadata is a **read** from SQL for local metadata for the remote model's id: 7.37s
- query pending mutations is a **read** from SQL for pending mutations not in process: 10.27s
- save model is a **read** (check if exists) and **write** (insert/update) to SQL: 39.87s
- save metadata is a **read** for local metadata, **read** (check if exists), and a **write** (insert/update) to SQL: 41.49s
- total 5 SQL reads and 2 writes, 99s
-  sync query API call has been tested before and takes around <1s (around 0.8s on average). there are 10 API calls (10k records/1k page)
- total elapsed time from the sample app `DataStore.start`, to `ModelSynced` event: 114s

*Check points:*


By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
